File: alligator/src/main/java/me/aartikov/alligator/navigators/ActivityNavigator.java
package me.aartikov.alligator.navigators;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import me.aartikov.alligator.Screen;
import me.aartikov.alligator.ScreenResult;
import me.aartikov.alligator.animations.AnimationData;
import me.aartikov.alligator.destinations.ActivityDestination;
import me.aartikov.alligator.exceptions.NavigationException;

public interface ActivityNavigator {

	void goForward(@NonNull Screen screen,
				   @NonNull ActivityDestination destination,
				   @Nullable AnimationData animationData) throws NavigationException;

	void replace(@NonNull Screen screen,
				 @NonNull ActivityDestination destination,
				 @Nullable AnimationData animationData) throws NavigationException;

	void reset(@NonNull Screen screen,
			   @NonNull ActivityDestination destination,
			   @Nullable AnimationData animationData) throws NavigationException;

	void goBack(@Nullable ScreenResult screenResult,
				@Nullable AnimationData animationData) throws NavigationException;

	void goBackTo(@NonNull Class<? extends Screen> screenClass,
				  @NonNull ActivityDestination destination,
				  @Nullable ScreenResult screenResult,
				  @Nullable AnimationData animationData) throws NavigationException;

}

File: alligator/src/main/java/me/aartikov/alligator/navigators/ActivityNavigator.kt
package me.aartikov.alligator.navigators

import me.aartikov.alligator.Screen
import me.aartikov.alligator.ScreenResult
import me.aartikov.alligator.animations.AnimationData
import me.aartikov.alligator.destinations.ActivityDestination
import me.aartikov.alligator.exceptions.NavigationException

interface ActivityNavigator {
    @Throws(NavigationException::class)
    fun goForward(
        screen: Screen,
        destination: ActivityDestination,
        animationData: AnimationData?
    )

    @Throws(NavigationException::class)
    fun replace(
        screen: Screen,
        destination: ActivityDestination,
        animationData: AnimationData?
    )

    @Throws(NavigationException::class)
    fun reset(
        screen: Screen,
        destination: ActivityDestination,
        animationData: AnimationData?
    )

    @Throws(NavigationException::class)
    fun goBack(
        screenResult: ScreenResult?,
        animationData: AnimationData?
    )

    @Throws(NavigationException::class)
    fun goBackTo(
        screenClass: Class<out Screen?>,
        destination: ActivityDestination,
        screenResult: ScreenResult?,
        animationData: AnimationData?
    )
}